// ng2-carouselamos/__ivy_ngcc__/fesm2015/ng2-carouselamos.js
import { EventEmitter, Component, Input, Output, ViewChild, NgModule } from '@angular/core';
import isEqual from 'lodash.isequal';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng2-carouselamos.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = ["list"];
const _c1 = ["ng2-carouselamos", ""];
function Ng2CarouselamosComponent_4_ng_template_0_Template(rf, ctx) { }
const _c2 = function (a0, a1, a2) { return { $implicit: a0, index: a1, selectedIndex: a2 }; };
function Ng2CarouselamosComponent_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, Ng2CarouselamosComponent_4_ng_template_0_Template, 0, 0, "ng-template", 6);
} if (rf & 2) {
    const item_r3 = ctx.$implicit;
    const i_r4 = ctx.index;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.$item)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction3(2, _c2, item_r3, i_r4, ctx_r1.childIndex));
} }
function Ng2CarouselamosComponent_div_5_button_1_ng_template_1_Template(rf, ctx) { }
function Ng2CarouselamosComponent_div_5_button_1_Template(rf, ctx) { if (rf & 1) {
    const _r10 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 9);
    ɵngcc0.ɵɵlistener("click", function Ng2CarouselamosComponent_div_5_button_1_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r10); const ctx_r9 = ɵngcc0.ɵɵnextContext(2); const _r0 = ɵngcc0.ɵɵreference(3); return ɵngcc0.ɵɵresetView(ctx_r9.scroll(false, _r0)); });
    ɵngcc0.ɵɵtemplate(1, Ng2CarouselamosComponent_div_5_button_1_ng_template_1_Template, 0, 0, "ng-template", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵproperty("disabled", ctx_r6.amount >= 0);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r6.$prev);
} }
function Ng2CarouselamosComponent_div_5_button_2_ng_template_1_Template(rf, ctx) { }
function Ng2CarouselamosComponent_div_5_button_2_Template(rf, ctx) { if (rf & 1) {
    const _r13 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 9);
    ɵngcc0.ɵɵlistener("click", function Ng2CarouselamosComponent_div_5_button_2_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r13); const ctx_r12 = ɵngcc0.ɵɵnextContext(2); const _r0 = ɵngcc0.ɵɵreference(3); return ɵngcc0.ɵɵresetView(ctx_r12.scroll(true, _r0)); });
    ɵngcc0.ɵɵtemplate(1, Ng2CarouselamosComponent_div_5_button_2_ng_template_1_Template, 0, 0, "ng-template", 10);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    const _r0 = ɵngcc0.ɵɵreference(3);
    ɵngcc0.ɵɵproperty("disabled", ctx_r7.amount <= -(_r0.scrollWidth - ctx_r7.width));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r7.$next);
} }
function Ng2CarouselamosComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 7);
    ɵngcc0.ɵɵtemplate(1, Ng2CarouselamosComponent_div_5_button_1_Template, 2, 2, "button", 8);
    ɵngcc0.ɵɵtemplate(2, Ng2CarouselamosComponent_div_5_button_2_Template, 2, 2, "button", 8);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.$prev);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r2.$next);
} }
function SelectedItemInterface() { }
if (false) {
    /** @type {?} */
    SelectedItemInterface.prototype.item;
    /** @type {?} */
    SelectedItemInterface.prototype.index;
    /** @type {?} */
    SelectedItemInterface.prototype.first;
}
/*
 *
 * @param() startAt - index of slide to render first. Default to 0.
 * @param() items - List of items to belong in carousel
 * @param() width - Size of window(view) to show
 * @param() $prev - Template for previous button
 * @param() $next - Template for next button
 * @param() $item - Template for the item
 */
class Ng2CarouselamosComponent {
    constructor() {
        this.childIndex = 0;
        this.amount = 0;
        this.startPress = 0;
        this.lastX = 0;
        this.startAt = 0;
        this.items = [];
        this.width = 500;
        this.onSelectedItem = new EventEmitter();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onMousedown(e) {
        if (e.which === 1) {
            this.startPress = e.clientX;
            this.lastX = this.amount;
        }
    }
    /**
     * @param {?} e
     * @return {?}
     */
    onTouchdown(e) {
        if (navigator.userAgent.indexOf("Android") >= 0)
            e.preventDefault();
        this.startPress = e.targetTouches[0].clientX;
        this.lastX = this.amount;
    }
    /**
     * @param {?} e
     * @param {?} maxWidth
     * @return {?}
     */
    onMousemove(e, maxWidth) {
        if (e.which === 1) {
            /** @type {?} */
            const amount = this.lastX - (this.startPress - e.clientX);
            if (amount > 0 || amount < -(maxWidth - this.width))
                return;
            this.amount = amount;
        }
    }
    /**
     * @param {?} e
     * @param {?} maxWidth
     * @return {?}
     */
    onTouchmove(e, maxWidth) {
        if (navigator.userAgent.indexOf("Android") >= 0)
            e.preventDefault();
        /** @type {?} */
        const amount = this.lastX - (this.startPress - e.targetTouches[0].clientX);
        if (amount > 0 || amount < -(maxWidth - this.width))
            return;
        this.amount = amount;
    }
    /**
     * @param {?} e
     * @param {?} elem
     * @return {?}
     */
    onMouseup(e, elem) {
        if (e.which === 1) {
            this.startPress = 0;
            this.snap(elem);
        }
    }
    /**
     * @param {?} e
     * @param {?} elem
     * @return {?}
     */
    onTouchup(e, elem) {
        if (navigator.userAgent.indexOf("Android") >= 0)
            e.preventDefault();
        this.startPress = 0;
        this.snap(elem);
    }
    /**
     * @param {?} elem
     * @return {?}
     */
    snap(elem) {
        /** @type {?} */
        let counter = 0;
        /** @type {?} */
        let lastVal = 0;
        for (let i = 0; i < this.items.length; i++) {
            /** @type {?} */
            const el = elem.children[i];
            /** @type {?} */
            const style = el.currentStyle || window.getComputedStyle(el);
            counter +=
                el.offsetWidth +
                    (parseFloat(style.marginLeft) + parseFloat(style.marginRight));
            if (this.amount <= lastVal && this.amount >= -counter) {
                this.amount = -lastVal;
                this.childIndex = i;
                this.onSelectedItem.emit({
                    item: this.items[this.childIndex],
                    index: this.childIndex,
                    first: false
                });
                return;
            }
            lastVal = counter;
        }
        return counter;
    }
    /**
     * @param {?} forward
     * @param {?} elem
     * @param {?=} qty
     * @return {?}
     */
    scroll(forward, elem, qty = 1) {
        this.childIndex += forward ? qty : -qty;
        this.onSelectedItem.emit({
            item: this.items[this.childIndex],
            index: this.childIndex,
            first: false
        });
        this.amount = -this.calcScroll(elem);
    }
    /**
     * @param {?} elem
     * @return {?}
     */
    calcScroll(elem) {
        /** @type {?} */
        let counter = 0;
        for (let i = this.childIndex - 1; i >= 0; i--) {
            /** @type {?} */
            const el = elem.children[i];
            /** @type {?} */
            const style = el.currentStyle || window.getComputedStyle(el);
            counter +=
                el.offsetWidth +
                    (parseFloat(style.marginLeft) + parseFloat(style.marginRight));
        }
        return counter;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.items &&
            !isEqual(changes.items.previousValue, changes.items.currentValue)) {
            if (changes.items.firstChange) {
                this.onSelectedItem.emit({
                    item: this.items[this.childIndex],
                    index: this.childIndex,
                    first: true
                });
            }
            this.amount = 0;
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.startPress = 1;
        this.scroll(true, this.list.nativeElement, this.startAt);
        setTimeout((/**
         * @return {?}
         */
        () => (this.startPress = 0)));
    }
}
Ng2CarouselamosComponent.ɵfac = function Ng2CarouselamosComponent_Factory(t) { return new (t || Ng2CarouselamosComponent)(); };
Ng2CarouselamosComponent.ɵcmp = /*@__PURE__*/ ɵngcc0.ɵɵdefineComponent({ type: Ng2CarouselamosComponent, selectors: [["", "ng2-carouselamos", ""]], viewQuery: function Ng2CarouselamosComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, 7);
    } if (rf & 2) {
        let _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.list = _t.first);
    } }, inputs: { startAt: "startAt", items: "items", width: "width", $prev: "$prev", $next: "$next", $item: "$item" }, outputs: { onSelectedItem: "onSelectedItem" }, features: [ɵngcc0.ɵɵNgOnChangesFeature], attrs: _c1, decls: 6, vars: 13, consts: [[1, "ng2-carouselamos-container"], [1, "ng2-carouselamos-wrapper", 3, "mousedown", "touchstart", "mousemove", "touchmove", "mouseup", "mouseleave", "touchend"], [1, "ng2-carouselamos"], ["list", ""], [4, "ngFor", "ngForOf"], ["class", "controls", 4, "ngIf"], [3, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "controls"], ["type", "button", 3, "disabled", "click", 4, "ngIf"], ["type", "button", 3, "disabled", "click"], [3, "ngTemplateOutlet"]], template: function Ng2CarouselamosComponent_Template(rf, ctx) { if (rf & 1) {
        const _r14 = ɵngcc0.ɵɵgetCurrentView();
        ɵngcc0.ɵɵelementStart(0, "div", 0)(1, "div", 1);
        ɵngcc0.ɵɵlistener("mousedown", function Ng2CarouselamosComponent_Template_div_mousedown_1_listener($event) { return ctx.onMousedown($event); })("touchstart", function Ng2CarouselamosComponent_Template_div_touchstart_1_listener($event) { return ctx.onTouchdown($event); })("mousemove", function Ng2CarouselamosComponent_Template_div_mousemove_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const _r0 = ɵngcc0.ɵɵreference(3); return ɵngcc0.ɵɵresetView(ctx.onMousemove($event, _r0.scrollWidth)); })("touchmove", function Ng2CarouselamosComponent_Template_div_touchmove_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const _r0 = ɵngcc0.ɵɵreference(3); return ɵngcc0.ɵɵresetView(ctx.onTouchmove($event, _r0.scrollWidth)); })("mouseup", function Ng2CarouselamosComponent_Template_div_mouseup_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const _r0 = ɵngcc0.ɵɵreference(3); return ɵngcc0.ɵɵresetView(ctx.onMouseup($event, _r0)); })("mouseleave", function Ng2CarouselamosComponent_Template_div_mouseleave_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const _r0 = ɵngcc0.ɵɵreference(3); return ɵngcc0.ɵɵresetView(ctx.onMouseup($event, _r0)); })("touchend", function Ng2CarouselamosComponent_Template_div_touchend_1_listener($event) { ɵngcc0.ɵɵrestoreView(_r14); const _r0 = ɵngcc0.ɵɵreference(3); return ɵngcc0.ɵɵresetView(ctx.onTouchup($event, _r0)); });
        ɵngcc0.ɵɵelementStart(2, "div", 2, 3);
        ɵngcc0.ɵɵtemplate(4, Ng2CarouselamosComponent_4_Template, 1, 6, null, 4);
        ɵngcc0.ɵɵelementEnd()();
        ɵngcc0.ɵɵtemplate(5, Ng2CarouselamosComponent_div_5_Template, 3, 2, "div", 5);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("width", ctx.width + "px");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵstyleProp("transition", ctx.startPress > 0 ? "none" : "-webkit-transform 1s")("webkit-transition", ctx.startPress > 0 ? "none" : "-webkit-transform 1s")("transform", "translateX(" + ctx.amount + "px)")("webkit-transform", "translateX(" + ctx.amount + "px)");
        ɵngcc0.ɵɵattribute("startPress", ctx.startPress);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.items);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.$prev || ctx.$next);
    } }, dependencies: [ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc1.NgTemplateOutlet], styles: [".ng2-carouselamos-container[_ngcontent-%COMP%]{position:relative;display:flex;justify-content:center}.ng2-carouselamos-wrapper[_ngcontent-%COMP%]{overflow:hidden}.ng2-carouselamos[_ngcontent-%COMP%]{display:flex;transition:transform 1s}.controls[_ngcontent-%COMP%]{pointer-events:none;position:absolute;display:flex;width:100%;justify-content:space-between;top:50%;left:0;transform:translateY(-50%)}.controls[_ngcontent-%COMP%]   button[_ngcontent-%COMP%]{pointer-events:all;background:0 0;border:0}"] });
Ng2CarouselamosComponent.propDecorators = {
    startAt: [{ type: Input }],
    items: [{ type: Input }],
    width: [{ type: Input }],
    $prev: [{ type: Input }],
    $next: [{ type: Input }],
    $item: [{ type: Input }],
    onSelectedItem: [{ type: Output }],
    list: [{ type: ViewChild, args: ["list", { static: true },] }]
};
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng2CarouselamosComponent, [{
        type: Component,
        args: [{ selector: "[ng2-carouselamos]", template: "<div class=\"ng2-carouselamos-container\">\n  <div\n    class=\"ng2-carouselamos-wrapper\"\n    [style.width]=\"width + 'px'\"\n    (mousedown)=\"onMousedown($event)\"\n    (touchstart)=\"onTouchdown($event)\"\n    (mousemove)=\"onMousemove($event, list.scrollWidth)\"\n    (touchmove)=\"onTouchmove($event, list.scrollWidth)\"\n    (mouseup)=\"onMouseup($event, list)\"\n    (mouseleave)=\"onMouseup($event, list)\"\n    (touchend)=\"onTouchup($event, list)\"\n  >\n    <div\n      class=\"ng2-carouselamos\"\n      [attr.startPress]=\"startPress\"\n      [style.transition]=\"startPress > 0 ? 'none' : '-webkit-transform 1s'\"\n      [style.webkitTransition]=\"\n        startPress > 0 ? 'none' : '-webkit-transform 1s'\n      \"\n      [style.transform]=\"'translateX(' + amount + 'px)'\"\n      [style.webkitTransform]=\"'translateX(' + amount + 'px)'\"\n      #list\n    >\n      <ng-template\n        *ngFor=\"let item of items; let i = index\"\n        [ngTemplateOutlet]=\"$item\"\n        [ngTemplateOutletContext]=\"{\n          $implicit: item,\n          index: i,\n          selectedIndex: childIndex\n        }\"\n      ></ng-template>\n    </div>\n  </div>\n  <div class=\"controls\" *ngIf=\"$prev || $next\">\n    <button\n      type=\"button\"\n      *ngIf=\"$prev\"\n      (click)=\"scroll(false, list)\"\n      [disabled]=\"amount >= 0\"\n    >\n      <ng-template [ngTemplateOutlet]=\"$prev\"></ng-template>\n    </button>\n    <button\n      type=\"button\"\n      *ngIf=\"$next\"\n      (click)=\"scroll(true, list)\"\n      [disabled]=\"amount <= -(list.scrollWidth - width)\"\n    >\n      <ng-template [ngTemplateOutlet]=\"$next\"></ng-template>\n    </button>\n  </div>\n</div>\n", styles: [".ng2-carouselamos-container{position:relative;display:flex;justify-content:center}.ng2-carouselamos-wrapper{overflow:hidden}.ng2-carouselamos{display:flex;transition:transform 1s}.controls{pointer-events:none;position:absolute;display:flex;width:100%;justify-content:space-between;top:50%;left:0;transform:translateY(-50%)}.controls button{pointer-events:all;background:0 0;border:0}"] }]
    }], function () { return []; }, { startAt: [{
            type: Input
        }], items: [{
            type: Input
        }], width: [{
            type: Input
        }], onSelectedItem: [{
            type: Output
        }], $prev: [{
            type: Input
        }], $next: [{
            type: Input
        }], $item: [{
            type: Input
        }], list: [{
            type: ViewChild,
            args: ["list", { static: true }]
        }] }); })();
if (false) {
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.childIndex;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.amount;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.startPress;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.lastX;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.startAt;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.items;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.width;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.$prev;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.$next;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.$item;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.onSelectedItem;
    /** @type {?} */
    Ng2CarouselamosComponent.prototype.list;
}

/**
 * @fileoverview added by tsickle
 * Generated from: lib/ng2-carouselamos.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class Ng2CarouselamosModule {
}
Ng2CarouselamosModule.ɵfac = function Ng2CarouselamosModule_Factory(t) { return new (t || Ng2CarouselamosModule)(); };
Ng2CarouselamosModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: Ng2CarouselamosModule });
Ng2CarouselamosModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [CommonModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(Ng2CarouselamosModule, [{
        type: NgModule,
        args: [{
                declarations: [Ng2CarouselamosComponent],
                imports: [CommonModule],
                exports: [Ng2CarouselamosComponent]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2CarouselamosModule, { declarations: function () { return [Ng2CarouselamosComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [Ng2CarouselamosComponent]; } }); })();

/**
 * @fileoverview added by tsickle
 * Generated from: public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * Generated from: ng2-carouselamos.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { Ng2CarouselamosComponent, Ng2CarouselamosModule };
